TMS-214: Vivi-Suche using Elasticsearch

witsch opened the pull request:
Siehe [TMS-214](https://zeit-online.atlassian.net/browse/TMS-214) & [TMS-240](https://zeit-online.atlassian.net/browse/TMS-240).

Erst deployen, wenn Autoren durchsuchbar re-indiziert sind ([ZON-4788](https://zeit-online.atlassian.net/browse/ZON-4788)), sonst tut das Type-Ahead nicht sinnvoll.

witsch committed:
TMS-240: Refactor (non-browser) tests so we can have several test modules.

witsch committed:
TMS-240: Add initial version of query builder for elasticsearch.

witsch committed:
TMS-240: Provide CLI scripts to query solr & elasticsearch.

  $ work/zope/bin/search-elastic --fulltext Foo

witsch committed:
TMS-240: Allow search script to report the query & results.

In addition search conditions are now specified using a <field>:<term>
notation:

  $ work/zope/bin/search-solr --verbose "authors:Klaus Mustermann"
  $ work/zope/bin/search-elastic --verbose "fulltext:Klaus Lage"

witsch committed:
TMS-240: Support date ranges via `from` and `until`.

witsch committed:
TMS-240: Allow the helper search script to include the payload.

  $ work/zope/bin/search-elastic --payload --fulltext Foo

witsch committed:
TMS-240: Map remaining field except for `show_news`.

Also note that arbitrary `filter_terms` are no longer supported (as
users would need to specify valid query JSON).

witsch committed:
TMS-240: Refactor to allow builders to specify the boolean "occurrence" type.

See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html

witsch committed:
TMS-240: Support the `show_news` search filter.

witsch committed:
TMS-240: Use a "match all" query if no filters were given.

witsch committed:
TMS-240: Only exclude news if `show_news` was actually set to `False`.

witsch committed:
TMS-240: Sometimes empty lists are passed meaning "match all".

witsch committed:
TMS-240: Fix types typo from 30dd8046... :)

witsch committed:
TMS-240: Simplify code as there were too few 'builder' cases anyway.

Also handling more than one field and the case where no new clause
was needed has become a bit awkward now...

witsch committed:
TMS-240: `fulltext` is sometimes passed as a positional argument.

witsch committed:
TMS-240: Add BBB import for `z.c.cp.automatic.SolrContentQuery`.

witsch committed:
TMS-240: Allow `from_` and `until` to be `None`.

witsch committed:
TMS-240: Fix `must_not` clauses.

witsch committed:
TMS-240: Move Solr search (module) aside so we can start testing things ttw.

Note that this will probably break most tests, which is okay for now...

witsch committed:
TMS-240: Add "drop-in" replacement `search` module using Elasticsearch.

witsch committed:
TMS-240: Use the score as the default order for now.

witsch committed:
TMS-240: Merge remote-tracking branch 'zeitonline/master' into 'TMS-240'

* zeitonline/master:
  MAINT: Remove unused images
  MAINT: Remove unused variables
  TMS-214: Update css naming after results filters have been removed
  TMS-214: Remove unused faceted/result-based filtering
  TMS-214: Replace unhelpful nn/yyyy with product in result display
  MAINT: jslint
  TMS-214: Remove base class after only other subclass (favourites) has been removed
  TMS-214: After favourites tab has been removed, remove the tab-bar altogether
  TMS-214: Remove unused "favourites" tab
  MAINT: Consistency
  TMS-214: Remove related section from search results
  TMS-214: Remove hit counter from search results
  TMS-214: Add "reset form" button

witsch committed:
TMS-240: Include payload by default.

It doesn't make much sense not to here, because we want to show
the search results (including some of their metadata).

witsch committed:
TMS-240: (Partially) map the data from Elasticsearch to the search results template.

witsch committed:
TMS-240: Fix proxy class and map remaining search result keys.

witsch committed:
TMS-240: Support multiple values and turn them into 'should' (read 'or') queries.

witsch committed:
TMS-240: Fix mapping for content type.

witsch committed:
TMS-240: Make it easier to debug queries against Elasticsearch.

witsch committed:
TMS-240: Fix base URL for 'uniqueId'.

witsch committed:
TMS-240: Explicitly specify result keys instead of using `include_payload`.

witsch committed:
TMS-240: Support 'date' and 'relevance' sort orders.

Please note that the score currently generated by Elasticsearch
seems rather useless. Perhaps we need to use `filter` for some
of the criteria.

witsch committed:
TMS-240: "Downgrade" what is really just debug output...

witsch committed:
TMS-240: Use `zeit_pool_content` index to also find unpublished content.

This needs to be cleaned up and moved into a utility and the product
config, though.

witsch committed:
TMS-240: Fix search for a specific "product".

The search form sends `product`, not `product_id`.

witsch committed:
TMS-240: The "keyword" search should also look for persons, locations etc.

witsch committed:
TMS-240: The "raw tag" field should search the content body.

witsch committed:
TMS-240: Only offer filters for types that have actually been indexed.

`advertisement` will be enabled once TMS-239 has been fixed.

witsch committed:
TMS-240: Use `p.t.supertitle` so "volume" content looks right.

They should read something like "Ausgabe 28/2018".

witsch committed:
TMS-240: Prefer structured data from `payload` over top-level fields.

This partially reverts 062f89bd, but `title` is still needed for
portrait boxes (as their `name` is otherwise not available and all
`teaser` and `body` fields do not get set).

witsch committed:
TMS-240: The preview URL received from Elasticsearch can be an empty string.

witsch committed:
TMS-240: Follow-up on 5d815006 to bring back the teaser text for image groups.

We should make sure all content is indexed with meaningful
`payload.body.*` fields as part of TMS-239. After that we can clean
up here, i.e. only use those fields.

witsch committed:
TMS-240: Style search reset "button" to be inside the search box.

This is a shameless plug of https://codepen.io/shidhincr/pen/ICLBD :)

witsch committed:
TMS-240: Move "Vorschau" and "CMS" links to the left.

witsch committed:
TMS-240: Don't "fold" the "types" & "extended" sections when resetting the form.

witsch committed:
TMS-240: Translate "no results" template.

The interface is already german anyway, or at least the other JSON
templates are.

witsch committed:
TMS-240: Configure (internal) search utility via product configuration.

This cleans up f2cbedb0.

witsch committed:
TMS-240: Fix PEP8 warning.

witsch committed:
TMS-240: Add BBB support for `suggest_query`.

However, the function is currently only used in one place
(`zeit.find.browser.autocomplete`) and should be "inlined"
there and removed completely afterwards.

witsch committed:
TMS-240: Add simple search test using a mocked elasticsearch result.

Also run the solr tests against the solr mock again.

wosc committed:
MAINT: Move testlayer to `testing` module, like all vivi packages do

wosc committed:
MAINT: Remove invalid namespace package declaration

wosc committed:
TMS-214: Bump version

wosc committed:
TMS-214: Introduce interfaces module, at least z.c.author needs to use this

wosc committed:
TMS-214: Remove solr code, fix test fixtures, tests pass now

wosc committed:
TMS-214: Update autocomplete result fields to ES names

wosc committed:
Update changelog

wosc committed:
TMS-214: Fold toplevel `search` function into the ICMSSearch utility

wosc committed:
MAINT: Clarify name

wosc reviewed (commented):
![wonderful](https://media.giphy.com/media/ddUrtZA5JZD6E/giphy.gif)

wosc committed:
TMS-214: Make autocomplete work with ES

NOTE that the payload.vivi.autocomplete field will only exist after the ZON-4788 reindex

witsch commented on src/zeit/find/tests/__init__.py:
```diff
@@ -1,61 +1 @@
-    name='WebdriverLayer', bases=(HTTP_LAYER,))
-SELENIUM_LAYER = gocept.selenium.WebdriverSeleneseLayer(
-    name='WebdriverSeleneseLayer', bases=(WD_LAYER,))
+# python package
```
Super, das wollte ich eigentlich auch noch machen... 👍

witsch commented on src/zeit/find/search.py:
```diff
@@ -20,10 +15,12 @@
+    def search(self, query, **kw):
+        query.setdefault('_source', DEFAULT_FIELDS)
+        kw.setdefault('rows', 50)
+        return super(Elasticsearch, self).search(query, **kw)
```
Ja okay, so ist's natürlich schicker! 😄

witsch reviewed (commented):
Die "Aufräumarbeiten" sehen gut aus! 👍  Hoffentlich kriegen wir dann [ZON-4788](https://zeit-online.atlassian.net/browse/ZON-4788) schnell an den Start...

witsch committed:
TMS-240: Follow up on b4af18f7 and remove last solr bits.

witsch committed:
TMS-240: Fix pyflakes warning.

witsch committed:
TMS-240: Fix up 55c45402 by lower-casing the search term.

witsch committed:
TMS-240: Try to use `payload.teaser` data first, but fall back on `payload.body`.

Also see https://github.com/ZeitOnline/zeit.retresco/pull/38/commits/6c14039b

wosc committed:
TMS-240: Fix up 8644498/55c45402 by clarifying intention in the test

wosc committed:
TMS-214: Use `filter` instead of `must` where possible, since ES then can cache those clauses

wosc committed:
ZON-4806: Extract `date_range` helper method

witsch commented on src/zeit/find/search.py:
```diff
@@ -86,13 +86,9 @@ def query(fulltext=None, **conditions):
         filter.append(dict(range={
-            'payload.document.last-semantic-change': lsc}))
+            'payload.document.last-semantic-change':
+            zeit.retresco.search.date_range(from_, until)}))
```
Oha! 😄

witsch committed:
TMS-240: Update a6c93106 to avoid using a reserved keyword.

witsch committed:
TMS-240: Restore 2caeee30, which was accidentally reverted in b4af18f7.

witsch commented:
[ZON-4788](https://zeit-online.atlassian.net/browse/ZON-4788) ist — bis auf die fehlenden "advertisement" Objekte — fertig, womit das type-ahead für die Autoren funktioniert und dieser PR gemerged werden kann.  Außerdem sind die Änderungen bereits auf vivi.devel.zeit.de getestet, so daß wir eine release für staging brauchen…
